apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/actions.ts

```diff
@@ -2,106 +2,128 @@
-  const session = await getServerSession(authOptions);
-  if (!session) throw new AuthorizationError("Not authorized");
-
-  const hasUserSurveyAccess = await canUserAccessSurvey(session.user.id, surveyId);
-  if (!hasUserSurveyAccess) throw new AuthorizationError("Not authorized");
-
-  const survey = await getSurvey(surveyId);
-  if (!survey?.id) {
-    throw new ResourceNotFoundError("Survey", surveyId);
-  }
-
-  return survey.resultShareKey;
-};
-
-export const deleteResultShareUrlAction = async (surveyId: string): Promise<void> => {
-  const session = await getServerSession(authOptions);
-  if (!session) throw new AuthorizationError("Not authorized");
-
-  const hasUserSurveyAccess = await canUserAccessSurvey(session.user.id, surveyId);
-  if (!hasUserSurveyAccess) throw new AuthorizationError("Not authorized");
-
-  const survey = await getSurvey(surveyId);
-  if (!survey?.id) {
-    throw new ResourceNotFoundError("Survey", surveyId);
-  }
-
-  await updateSurvey({ ...survey, resultShareKey: null });
-};
-
-export const getEmailHtmlAction = async (surveyId: string) => {
-  const session = await getServerSession(authOptions);
-  if (!session) throw new AuthorizationError("Not authorized");
-
-  const hasUserSurveyAccess = await canUserAccessSurvey(session.user.id, surveyId);
-  if (!hasUserSurveyAccess) throw new AuthorizationError("Not authorized");
-
-  return await getEmailTemplateHtml(surveyId);
-};
+import { ResourceNotFoundError } from "@formbricks/types/errors";
+
+const ZSendEmbedSurveyPreviewEmailAction = z.object({
+  surveyId: z.string(),
+});
+
+export const sendEmbedSurveyPreviewEmailAction = authenticatedActionClient
+  .schema(ZSendEmbedSurveyPreviewEmailAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "read"],
+    });
+
+    const survey = await getSurvey(parsedInput.surveyId);
+    if (!survey) {
+      throw new ResourceNotFoundError("Survey", parsedInput.surveyId);
+    }
+
+    const rawEmailHtml = await getEmailTemplateHtml(parsedInput.surveyId);
+    const emailHtml = rawEmailHtml
+      .replaceAll("?preview=true&amp;", "?")
+      .replaceAll("?preview=true&;", "?")
+      .replaceAll("?preview=true", "");
+
+    return await sendEmbedSurveyPreviewEmail(
+      ctx.user.email,
+      "Formbricks Email Survey Preview",
+      emailHtml,
+      survey.environmentId
+    );
+  });
+
+const ZGenerateResultShareUrlAction = z.object({
+  surveyId: z.string(),
+});
+
+export const generateResultShareUrlAction = authenticatedActionClient
+  .schema(ZGenerateResultShareUrlAction)
+  .action(async ({ ctx, parsedInput }) => {
+    await checkAuthorization({
+      userId: ctx.user.id,
+      organizationId: await getOrganizationIdFromSurveyId(parsedInput.surveyId),
+      rules: ["survey", "update"],
```

Comment:
Let's make two auth checks here: `"response", "update"` + `"survey", "update"`
It's not perfect, but given users who have the right to update the survey and the responses also the authority to distribute the results and update the survey object, seems like the closest fit.